Feature/bridge bot

luis-nexus opened the pull request:
# BridgeBot - 주기적 토큰 브릿지 자동화 컨트랙트

## 개요

BridgeBot은 설정된 주기에 따라 자동으로 토큰을 브릿지하는 스마트 컨트랙트입니다. 사용자가 설정한 간격에 따라 지정된 수량의 토큰을 대상 체인으로 자동 전송합니다.

## 주요 기능

### 🔄 주기적 브릿지 실행
- 설정된 시간 간격(초 단위)에 따라 자동 브릿지 실행
- 예: 86400초 (24시간) 간격으로 매일 실행
- 스마트한 시간 체크: `현재시간 - (현재시간 % 간격)`으로 중복 실행 방지

### ⚙️ 유연한 설정 관리
- 토큰 주소, 수량, 받는 주소, 대상 체인 ID 설정 가능
- 오너 권한으로 설정 추가/수정/활성화/비활성화
- 여러 브릿지 설정을 동시에 운영 가능

### 🔐 보안 기능
- 오너만 설정 변경 및 출금 가능
- 누구나 브릿지 실행 가능 (실행 권한 분리)
- 재진입 공격 방지

### 💰 자금 관리
- ERC20 토큰 및 네이티브 토큰 지원
- 오너 권한으로 언제든지 출금 가능
- 부분/전체 출금 기능

luis-nexus committed:
impl bridge bot

luis-nexus committed:
add binding/go

luis-nexus committed:
주석 변경

luis-nexus committed:
accessControl

luis-nexus committed:
오픈제플린 상속받아 구현

luis-nexus committed:
feat: Add BridgeBot contract for automated periodic token bridging

- Implement BridgeBot contract with AccessControlDefaultAdminRules
- Add EXECUTOR_ROLE for bridge execution permission control
- Support dynamic bridge configuration with interval-based execution
- Implement owner functions for config management and fund withdrawal
- Fix security issues: reset lastExecuted on config update, use exact approval amounts
- Remove executeBridgeBatch and getExecutableConfigs functions
- Add comprehensive test suite for BridgeBot
- Update Go bindings for all contracts

zayden-nexus commented on src/BridgeBot.sol:
```diff
@@ -0,0 +1,384 @@
+        bridgeConfigs[configId].recipient = recipient;
+        bridgeConfigs[configId].toChainID = toChainID;
+        bridgeConfigs[configId].interval = interval;
+        bridgeConfigs[configId].lastExecuted = 0; // Reset execution timer
```
초기화여부를 입력받아도 좋을듯 합니다.
현재는 interval 등만 수정하고싶을때에도 강제로 0

zayden-nexus commented on src/BridgeBot.sol:
```diff
@@ -0,0 +1,384 @@
+            if (currentAllowance < totalRequired) {
+                // Reset allowance to 0 first if needed (for some tokens like USDT)
+                if (currentAllowance > 0) token.approve(address(bridge), 0);
+                token.approve(address(bridge), totalRequired);
```
approve 자체가 초기화 후 재설정하는것 처럼 덮어씌워짐.
+ SafeERC20.forceApprove 사용 추천

luis-nexus replied:
의견 감사합니다. 반영했습니다.

zayden-nexus commented on src/BridgeBot.sol:
```diff
@@ -0,0 +1,384 @@
+     * @param interval Execution interval in seconds
+     */
+    function updateBridgeConfig(uint configId, address tokenAddress, address recipient, uint toChainID, uint interval)
+        external
```
일반적인 운영중 update나 추가될 가능성이 있는 함수의 권한은 EDITOR_ROLE 같은거로 구분해두는것을 추천.
(constructor에서 defaultAdmin도 EDITOR_ROLE에 등록)

luis-nexus replied:
의견 감사합니다. 반영했습니다.

zayden-nexus commented on src/BridgeBot.sol:
```diff
@@ -0,0 +1,384 @@
+     * @param configId Configuration ID
+     * @param amount Amount to bridge
+     */
+    function executeBridge(uint configId, uint amount) external onlyRole(EXECUTOR_ROLE) nonReentrant {
```
실행하는 backend 봇에서 amount를 결정하는건가요?

luis-nexus replied:
네 금액이 매번 바뀔수 있습니다.

luis-nexus committed:
accept zayden pr

luis-nexus committed:
pr 의견 반영

luis-nexus committed:
refactor: improve BridgeBot contract with role management and code optimizations

luis-nexus committed:
impl script

luis-nexus committed:
refactor: optimize token approval and deployment script

- Change token approval to max allowance for gas efficiency
- Only re-approve when current allowance is insufficient
- Refactor deployment script to use function parameters instead of env vars
- Simplify deployment configuration with direct parameter passing

luis-nexus committed:
remove array for config

luis-nexus committed:
update bind go

luis-nexus committed:
remove duplicated variable

luis-nexus committed:
update go bind

luis-nexus committed:
remove unusable code

zayden-nexus commented on src/BridgeBot.sol:
```diff
@@ -0,0 +1,351 @@
+        if (balance > 0) {
+            (bool success,) = to.call{value: balance}("");
+            require(success, "Transfer failed");
+            emit NativeWithdrawn(to, balance);
```
Const.NATIVE_TOKEN 을 이미 사용중이라 
TokenWithdrawn, NativeWithdrawn 두 이벤트를 Withdrawn 로 통일하고 token주소 똑같이 넣어줘도 좋을것 같습니다. (나중에 모니터링같은 시스템 만들때 한가지이벤트로 처리)

luis-nexus replied:
구현했습니다!@

luis-nexus committed:
커밋 버전 반영

zayden-nexus reviewed (approved):
LGTM
